0002-ether-htlc-dynamic-final-addresses.md

```diff
@@ -0,0 +1,121 @@
+The nature of Bitcoin allows the COMIT node to handle all the keys involved in the HTLC and simply ask the user for where they want the funds when the time comes.
+To me this is good design and what we want in COMIT even outside of atomic swap protocols:
+
+1. The users negotiate a protocol with no keys required from the users' wallets
+2. The protocol execution starts with users funding the initial contracts from their wallets
+3. The COMIT nodes execute the protocol (via the user's instruction and/or automatically).
+4. The protocol execution completes by the users' getting the final assets they own into their wallets.
+
+As a rule of thumb, the COMIT node should own the private keys of whatever identities are embedded in the contracts on the blockchains.
+
+Since Ethereum is flexible we can restructure it to mimic how Bitcoin works and fit it into the above flow.
+
+## Research
+
+Bitcoin requires you witness the HTLC output with the secret and a signature (or a just the refund signature after a timeout).
+The transaction that has the witness will specify where to redeem the funds in its output.
+
+To follow the same pattern in Ethereum, our redeem transaction will need in its data:
+
+1. The secret
+2. The address to give the money to (the final redeem address)
+3. a signature under the redeem identity from the COMIT node
+
+The refund would be the same without (1)
+
+### The Final Address
+
+The decision we have to make here is whether the final address is:
+
+1. Provided in the data
+2. The sender address
+
+(2) is nice because it means that there is no way to mess it up and send a transaction to an address you don't own. It also saves a bit of gas.
+
+On the other hand, that might be a feature.
+You may want to send the asset in the HTLC to a contract or something, so I'm leaning towards (1).
+
+
+### The signature
+
+The key things when deciding what to sign are:
+
+1. It must not be possible to front run the transaction, learn the secret, and then create another transaction that redeems to a different address.
+2. It must not be possible to re-use the signature on a different chain or a different contract on the same chain.
+
+To ensure (1) we need to sign the final address (redeem or refund) with the private key of the corresponding identity.
+To ensure (2) we should sign the HTLC's contract address, the network id and chain id.
+
+So I propose that we sign the following messages for each outcome:
+
+- Redeem: `network_id || chain_id || htlc_contract_address || final_redeem_address`
+- Refund: `network_id || chain_id || htlc_contract_address || final_refund_address`
+
+We have to hash the message before signing it.
+In this case we can choose any 256 bit output hash function because we control both the singing code (in COMIT) and the verifier (in the contract).
+Our choices are `SHA-256` and `Keccak-256`.
```

Comment:
The reason why I mentioned sha-256 was actually because I wasn't sure of which hash function `ecrecover` would accept. All I could find was that it's a hashed messaged :) Did you verify that? Otherwise we can just drop sha-256.

Reply:
`ecrecover` doesn't care. It just takes 32 bytes (within the secp256k1 curve order I think). Keccak-256 it is then.